### PR TITLE
Preserve Data

### PR DESCRIPTION
**Pull request by JustoRago**
In this pull request, we have done the following:

- Saved all the data (books, people, and rentals) into JSON files. One per type of data storing and named accordingly.
- On startup of the app we load all data from files.  We have handled the case when one or more files are missing.

Thanks for your review!

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -6,12 +7,29 @@ def initialize(title, author)
...
+  end
+
+  def tojson(temp_hash)
+    if File.zero?('books.json')
+      File.write('books.json', JSON.pretty_generate([temp_hash]))
+    else
+      data_from_json = JSON.parse(File.read('books.json'))
+      data_from_json = [data_from_json] if data_from_json.class != Array
+      File.write('books.json', JSON.pretty_generate(data_from_json << temp_hash))
+    end
```
Comment (omar-muhamad) [inline book.rb]: - [x] The book data are saved directly after creating a new book. As per this [requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_preserve_data.md#:~:text=On%20exit%20from%20your%20app%20save%20all%20the%20data%20(books%2C%20people%2C%20and%20rentals)%20into%20JSON%20files.), Please save this data on exit only.

File: teacher.rb [hunk 2]
```diff
@@ -11,4 +11,28 @@ def initialize(age, name, specialization, id: Random.rand(1..1000))
...
+
+  def tojson(temp_hash)
+    if File.zero?('persons.json')
+      File.write('persons.json', JSON.pretty_generate([temp_hash]))
+    else
+      data_from_json = JSON.parse(File.read('persons.json'))
+      data_from_json = [data_from_json] if data_from_json.class != Array
+      File.write('persons.json', JSON.pretty_generate(data_from_json << temp_hash))
+    end
+  end
```
Comment (omar-muhamad) [inline teacher.rb]: - [x] Same here, As per this [requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_preserve_data.md#:~:text=On%20exit%20from%20your%20app%20save%20all%20the%20data%20(books%2C%20people%2C%20and%20rentals)%20into%20JSON%20files.), Please save the person data on exit only.

File: student.rb [hunk 3]
```diff
@@ -17,4 +17,28 @@ def add_classroom(classroom)
...
+
+  def tojson(temp_hash)
+    if File.zero?('persons.json')
+      File.write('persons.json', JSON.pretty_generate([temp_hash]))
+    else
+      data_from_json = JSON.parse(File.read('persons.json'))
+      data_from_json = [data_from_json] if data_from_json.class != Array
+      File.write('persons.json', JSON.pretty_generate(data_from_json << temp_hash))
+    end
+  end
```
Comment (omar-muhamad) [inline student.rb]: - [x] Same here, As per this [requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_preserve_data.md#:~:text=On%20exit%20from%20your%20app%20save%20all%20the%20data%20(books%2C%20people%2C%20and%20rentals)%20into%20JSON%20files.), Please save the person data on exit only.

File: rental.rb [hunk 4]
```diff
@@ -9,8 +9,64 @@ def initialize(date, book, person)
...
+        'person' => person_hash
+      }
+    }
+    if File.zero?('rentals.json')
+      File.write('rentals.json', JSON.pretty_generate([temp_hash]))
+    else
+      data_from_json = JSON.parse(File.read('rentals.json'))
+      data_from_json = [data_from_json] if data_from_json.class != Array
+      File.write('rentals.json', JSON.pretty_generate(data_from_json << temp_hash))
+    end
```
Comment (omar-muhamad) [inline rental.rb]: - [x] Same here, As per this [requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_preserve_data.md#:~:text=On%20exit%20from%20your%20app%20save%20all%20the%20data%20(books%2C%20people%2C%20and%20rentals)%20into%20JSON%20files.), Please save the rentals data on exit only.

- [x] There is an error that occurred when I try to add a new rental, So please fix that.

image

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @asdt560 and @shakerAbuDrais,

Nice job, you are almost there.



## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used the correct workflow.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by frank1738 — APPROVED** [review]
Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

✅ Correct use of Gitflow
✅ No linter error
✅ App is working as expected 👍 

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**